speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/FileDataStorage.cpp

```diff
@@ -98,64 +121,307 @@ void FileDataStorage::Shutdown()
+template <typename TNumber>
+void FileDataStorage::AddToArray(TNumber p_values[], TNumber p_value, unsigned long p_compressionStep) const
+{
+    int p_placeInArray = static_cast<int>(p_compressionStep % static_cast<unsigned long>(m_compressionRate));
+    p_values[p_placeInArray] = p_value;
+}
+
+/// @brief Get the median of the current compression step using Randomized QuickSelect
+/// @param p_values Array with values from the current compression step
+/// @return The median of the past time steps for a variable
+float FileDataStorage::GetMedian(float* p_values)
+{
+    float a = -1, b = -1;
+    GetMedianUtil(p_values, 0, m_compressionRate - 1, static_cast<int>(m_compressionRate / 2), a, b);
+    return b;
+}
+
+/// @brief Utility function to recursively get the median
+/// @param p_values Array with values from the current compression step
+/// @param p_start start of the current section of the array
+/// @param p_end end of the current section of the array
+/// @param p_middle middle of the current section of the array
+/// @param p_a number before the median of p_values
+/// @param p_b median of p_values
+void FileDataStorage::GetMedianUtil(float* p_values, int p_start, int p_end, int p_middle, float& p_a, float& p_b)
+{
+    if (p_start <= p_end)
     {
-        WRITE_STRING(m_outputStream, "LightsDecision");
-        WRITE_VAR(m_outputStream, p_decisions.GetLights());
+        int partitionIndex = RandomPartition(p_values, p_start, p_end);
+
+        if (partitionIndex == p_middle)
+        {
+            p_b = p_values[partitionIndex];
+            if (p_a != -1)
+            {
+                return;
+            }
+        }
+        else if (partitionIndex == p_middle - 1)
+        {
+            p_a = p_values[partitionIndex];
+            if (p_b != -1)
+            {
+                return;
+            }
+        }
+
+        if (partitionIndex >= p_middle)
+        {
+            GetMedianUtil(p_values, p_start, partitionIndex - 1, p_middle, p_a, p_b);
+        }
+        else
+        {
+            GetMedianUtil(p_values, partitionIndex + 1, p_end, p_middle, p_a, p_b);
+        }
     }
-    WRITE_STRING(m_outputStream, "NONE");
+}
+
+/// @brief partitions p_values around a randomly picked element
+/// @param p_values Array with values from the current compression step
+/// @param p_start start of the partition
+/// @param p_end end of the partition
+int FileDataStorage::RandomPartition(float* p_values, int p_start, int p_end)
+{
+    int pivot = m_random.NextInt(p_start, p_end + 1);
+    std::swap(p_values[pivot], p_values[p_end]);
+    return Partition(p_values, p_start, p_end);
+}
+
+/// @brief returns the correct position of the pivot element
+/// @param p_values Array with values from the current compression step
+/// @param p_start start of the partition
+/// @param p_end end of the partition
+int FileDataStorage::Partition(float* p_values, int p_start, int p_end)
+{
+    float lst = p_values[p_end];
```

Comment:
Please add a comment what lst stands for (I am assuming it stands for something)